lib/Monads/trace/Trace_Prefix_Closed.thy

```diff
@@ -0,0 +1,301 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ * Copyright 2020, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *)
+
+theory Trace_Prefix_Closed
+  imports
+    Trace_No_Trace
+    Trace_Monad_Equations
+    Eisbach_Tools.Rule_By_Method
+begin
+
+subsection "Prefix Closed"
+
+definition prefix_closed :: "('s, 'a) tmonad \<Rightarrow> bool" where
+  "prefix_closed f = (\<forall>s. \<forall>x xs. (x # xs) \<in> fst ` f s \<longrightarrow> (xs, Incomplete) \<in> f s)"
+
+lemmas prefix_closedD1 = prefix_closed_def[THEN iffD1, rule_format]
+
+lemma in_fst_snd_image_eq:
+  "x \<in> fst ` S = (\<exists>y. (x, y) \<in> S)"
+  "y \<in> snd ` S = (\<exists>x. (x, y) \<in> S)"
+  by (auto elim: image_eqI[rotated])
+
+lemma in_fst_snd_image:
+  "(x, y) \<in> S \<Longrightarrow> x \<in> fst ` S"
+  "(x, y) \<in> S \<Longrightarrow> y \<in> snd ` S"
+  by (auto simp: in_fst_snd_image_eq)
+
+lemmas prefix_closedD = prefix_closedD1[OF _ in_fst_snd_image(1)]
+
+lemma no_trace_prefix_closed:
+  "no_trace f \<Longrightarrow> prefix_closed f"
+  by (auto simp add: prefix_closed_def dest: no_trace_emp)
+
+
+subsection \<open>Set up for @{method wp}\<close>
+
+lemma prefix_closed_is_triple[wp_trip]:
+  "prefix_closed f = triple_judgement \<top> f (\<lambda>() f. id prefix_closed f)"
+  by (simp add: triple_judgement_def split: unit.split)
+
+
+subsection \<open>Rules\<close>
+
+(*
+  Collect generic prefix_closed lemmas here:
+   - naming convention is prefix_closed_NAME.
+   - add lemmas with assumptions to [prefix_closed_cond] set
+   - add lemmas without assumption to [prefix_closed_terminal] set
+*)
+
+named_theorems prefix_closed_terminal
+named_theorems prefix_closed_cond
+
+lemmas [prefix_closed_terminal] = no_trace_terminal[THEN no_trace_prefix_closed]
+
+lemma prefix_closed_bind[wp_split]:
+  "\<lbrakk>prefix_closed f; \<And>x. prefix_closed (g x)\<rbrakk> \<Longrightarrow> prefix_closed (f >>= g)"
+  apply (subst prefix_closed_def, clarsimp simp: bind_def)
+  apply (auto simp: Cons_eq_append_conv split: tmres.split_asm
+             dest!: prefix_closedD[rotated];
+         fastforce elim: rev_bexI)
```

Comment:
Sometimes it helps to give auto deeper depth parameters e.g. `auto (6,3) simp: ..`. Not actually sure if that is less dodgy than `auto ..; fastforce ..` or more, because people tend to not know these, but I wanted to put it out there. (Also not sure if it'll actually help -- the first number is the depth given to `blast` inside `auto`, default is 4, and the second is another depth parameter for another tactic that I don't actually know (default = 2)).

Reply:
(The other one to try, which you might already have done anyway is `force` instead of `auto`)

Reply:
I'll give both of those a go, it would be nice to avoid `auto ..; fastforce ..`

Reply:
It turns out that `auto 4 4 ..` does work, so I guess that just leaves the question of which is less dodgy.

Reply:
I would say `auto 4 4` is marginally less dodgy than `auto; fastforce`, but it's a toss-up. Happy to leave it up to your taste, you're the author :-)